pandas/core/algorithms.py

```diff
@@ -181,6 +182,20 @@ def _ensure_data(values: ArrayLike) -> np.ndarray:
     return ensure_object(values)
 
 
+@overload
+def _reconstruct_data(
+    values: ExtensionArray, dtype: DtypeObj, original: AnyArrayLike
+) -> ExtensionArray:
+    ...
+
+
+@overload
+def _reconstruct_data(
+    values: np.ndarray, dtype: DtypeObj, original: AnyArrayLike
+) -> np.ndarray:
+    ...
+
+
 def _reconstruct_data(
     values: ArrayLike, dtype: DtypeObj, original: AnyArrayLike
 ) -> ArrayLike:
```

Comment:
Just a suggestion here - not sure if it's worth doing, but if you did:
```python
ArrayLikeT = TypeVar("ArrayLikeT", ExtensionArray, np.ndarray)

def _reconstruct_data(values: ArrayLikeT, dtype: DtypeObj, origin: AnyArrayLike) -> ArrayLikeT:
```
couldn't you then avoid the `overload` ?

Reply:
I will do that - I have a feeling such a TypeVar might also be useful in other places :)